jetty-server/src/main/java/org/eclipse/jetty/server/HttpChannel.java

```diff
@@ -481,33 +481,44 @@ public boolean handle()
-                            boolean hasContent = !(_request.isHead() ||
-                                    HttpMethod.CONNECT.is(_request.getMethod()) && status == HttpStatus.OK_200 ||
-                                    HttpStatus.isInformational(status) ||
-                                    status == HttpStatus.NO_CONTENT_204 ||
-                                    status == HttpStatus.NOT_MODIFIED_304);
-                            if (hasContent && !_response.isContentComplete(_response.getHttpOutput().getWritten()))
+                            try
                             {
-                                if (isCommitted())
-                                    abort(new IOException("insufficient content written"));
+                                if (!_response.isCommitted() && !_request.isHandled())
+                                {
+                                    _response.sendError(HttpStatus.NOT_FOUND_404);
+                                }
                                 else
-                                    _response.sendError(HttpStatus.INTERNAL_SERVER_ERROR_500,"insufficient content written");
+                                {
+                                    // RFC 7230, section 3.3.
+                                    int status = _response.getStatus();
+                                    boolean hasContent = !(_request.isHead() ||
+                                        HttpMethod.CONNECT.is(_request.getMethod()) && status == HttpStatus.OK_200 ||
+                                        HttpStatus.isInformational(status) ||
+                                        status == HttpStatus.NO_CONTENT_204 ||
+                                        status == HttpStatus.NOT_MODIFIED_304);
+                                    if (hasContent && !_response.isContentComplete(_response.getHttpOutput().getWritten()))
+                                    {
+                                        if (isCommitted())
+                                            abort(new IOException("insufficient content written"));
+                                        else
+                                            _response.sendError(HttpStatus.INTERNAL_SERVER_ERROR_500, "insufficient content written");
+                                    }
+                                }
+                                _response.closeOutput();
+
+                            }
+                            finally
+                            {
+                                _request.setHandled(true);
+                                _state.onComplete();
                             }
                         }
-                        _response.closeOutput();
-                        _request.setHandled(true);
-
-                        _state.onComplete();
-
-                        onCompleted();
+                        finally
+                        {
+                            onCompleted();
+                        }
```

Comment:
Why do we need 2 nested try-finally?

If `_state.onComplete()` throws from the inner `finally` block, then we are possibly hiding an exception thrown by the innermost code.

I think it should be guaranteed that `_state.onComplete()` does not throw and we can merge the 2 nested try/finally into one.

Reply:
Done